File: javascript/list-ops/list-ops.js
export class List {
  constructor(items = []) {
      this._data = items;
  }

  get values() {
      return [...this._data];
  }

  append(otherList) {
      return new List([...this._data, ...otherList.values]);
  }

  concat(lists) {
      let r = this.values;
      for (let l of lists.values) {
          r.push(...l.values);
      }

      return new List(r);
  }

  filter(predicate) {
      let r = [];
      for (let item of this.values) {
          if (predicate(item)) {
              r.push(item);
          }
      }

      return new List(r);
  }

  map(transform) {
      let r = [];
      for (let item of this.values) {
          r.push(transform(item))
      }

      return new List(r);
  }

  length() {
      return this._data.length;
  }

  foldl(reducer, initialValue) {
      let acc = initialValue;
      for (let item of this.values) {
          acc = reducer(acc, item);
      }
      return acc;
  }

  foldr(reducer, initialValue) {
      let acc = initialValue;
      let items = this.values;
      while (items.length > 0) {
          acc = reducer(acc, items.pop());
      }
      return acc;
  }

  reverse() { 
      let r = [];
      let items = this.values;
      while (items.length > 0) {
          r.push(items.pop());
      }
      return new List(r);
  }
}

File: python/list-ops/list_ops.py
def append(list1, list2):
    if not list1:
        return list2

    head, *tail = list1

    return [head] + append(tail, list2)


def concat(lists):
    if not lists:
        return []

    head, *tail = lists

    return append(head, concat(tail))


def filter(function, list):

    if not list:
        return []

    head, *tail = list

    if function(head):
        return [head] + filter(function, tail)

    return filter(function, tail)


def length(list):

    if not list:
        return 0

    _, *tail = list

    return 1 + length(tail)


def map(function, list):

    if not list:
        return []

    head, *tail = list

    return [function(head)] + map(function, tail)


def foldl(function, list, initial):

    if not list:
        return initial

    head, *tail = list

    return foldl(function, tail, function(initial, head))


def foldr(function, list, initial):

    if not list:
        return initial

    head, *tail = list

    return function(foldr(function, tail, initial), head)


def reverse(list):

    if not list:
        return []

    head, *tail = list

    return reverse(tail) + [head]
